[2기-P] 곽동운 SpringBoot-Board-JPA 미션 제출합니다

midasWorld opened the pull request:
# 🛠 작업 단위

- 🔖 미션

  - [x] User, Post 엔티티 구성
  - [x] 게시글 조회 (페이징)
  - [x] 게시글 단건조회, 작성, 수정
  - [ ] REST-DOCS

- 🚀 추가사항

  - [ ] Bean Validation 적용
  - [ ] 회원가입, 로그인 기능
  - [ ] Test 코드 작성
  - [ ] 좋아요 기능, 댓글 기능

# 📌 PR 포인트?

1. [Paging 부분 🗂](#1-paging-부분-🗂)
2. [QueryDSL 사용 이유 👀..](#2-querydsl-사용-이유-👀)
3. [Service단에서 유효성 검사 🔎](#3-service단에서-유효성-검사-🔎)
4. [Post Request Dto → [유효성검사 제외] & [Setter 사용 → Create, Update 두개 용도로 합침]](#4-post-request-dto-→-유효성검사-제외--setter-사용-→-create-update-두개-용도로-합침)
5. [Service → Entity 반환 🤔](#5-service-→-entity-반환-🤔)
6. [그 외](#그-외)

### 1. Paging 부분 🗂

Commit: [bb6e70a](https://github.com/midasWorld/springboot-board-jpa/commit/bb6e70a7bfc187f3e2b2844b062634fe92abcbbf)

PageRequest와 Resolver를 구현 및 등록해서 사용했습니다.
Page 그대로 사용하지 않은 이유는 count 쿼리 때문이었고, 그를 위해서 count 메서드를 따로 분리시켰습니다.
(사실 이 부분은 튜닝의 영역이고 구축단계에서 필요한가? 라는 생각도 드네요.)

일단은 프론트단에서 대충이라도 페이징을 좀 구현을 해보면서 필요한 부분들을 찾아볼까 하고 있습니다.

최종적으로는 Cursor 방식으로 변경을 하는게 목표입니다.

---

### 2. QueryDSL 사용 이유 👀..

시작은 1번의 Offset과 Limit를 좀 더 직관적으로 코드로 구현할 수 있어서 였고..
추후에 조회용 DTO를 post 도메인에 만들어서, 필요한 컬럼들만 조회를 하고자 사용하게 되었습니다.

---

### 3. Service단에서 유효성 검사 🔎

일단 저는 Model > Service > Controller 순으로 유효성 검사의 중요도 순이라고 생각하고 있습니다.
Model, Service단에는 넣었고, Controller단에서 넣어야 하는지 고민이었고, 아직 고민하고 있습니다. 😭

Controller단에도 있어야 된다고 생각하면서도 Service단 메서드 들어가자마자 검사를 해서 중복된다라는 느낌을 계속 받아서 Reqeust DTO 부분에 유효성 검사를 제외했습니다.

---

### 4. Post Request Dto → [유효성검사 제외] & [Setter 사용 → Create, Update 두개 용도로 합침].

Commit: [828571b](https://github.com/midasWorld/springboot-board-jpa/commit/828571bdcbcc2fa01f82f38a0cc146382ff70089)

3번과 이어지는 부분인데요. 유효성 검사 부분을 먼저 제외 했습니다.

일단 PostRequest에 Setter를 추가해서 update의 경우에는 writerId를 함께 요청하지 않아도 받을 수 있도록 했습니다.
Service단에서 유효성 검사를 하고 있고, Controller의 경우에는 데이터를 그대로 받아서 Service단으로 요청한다라는 의미로 사용해도 되지 않을까 싶어서... Setter를 추가해서 느슨하게 사용하였습니다.

---

## 5. Service → Entity 반환 🤔

DTO로 변환해서 반환되도록 사용을 해봤는데, 어떤식으로 해야 될지 감이 잡히지 않네요.. 😓
그래서 지금 방식으로 해서 문제점들도 좀 부딪혀보고, 다른 분들 코드도 보면서 공부를 해보려고 합니다... 👶..
(일단 open-in-view 속성을 false 설정하였습니다.)

---

### 그 외

- CreatedBy 는 로그인 기능 구현과 연계해서 하기 위해, nullable = true로 해서 null 값이 일단 들어가도록 하였습니다.
- Indent 다른거 하다가 2로 해버렸습니다... 이거 다음 PR때 적용해서 진행될 수 있도록 하겠습니다. 😭

midasWorld committed:
init: 게시판 프로젝트 생성

midasWorld committed:
setting: application.yml - DB 설정 정보 추가

midasWorld committed:
setting: application.yml - jpa open-in-view: false 설정 추가

midasWorld committed:
feat: 회원 목록 조회, ID 조회 기능 구현
- BaseEntity, User 엔티티/레파지토리 생성
- UserService 생성
- AuditConfig 생성 - CreatedBy 등록용
- ApiResponse, ApiError 생성 - Controller 반환용
- UserRestController 생성
- data.sql - User 샘플 데이터 등록

midasWorld committed:
feat: GlobalExceptionHandler 생성 - 전역 예외 처리용

midasWorld committed:
setting: pom.xml - querydsl 라이브러리 추가

midasWorld committed:
feat: Post 페이징 조회 구현 - count 쿼리는 따로 분리
- Pageable 생성 및 Handler 구현 및 설정
- Post 엔티티, PostRepository 생성 및 count 메서드 구현
- PostQueryRepository 생성 (Post 페이징 조회용) - offset, limit 조회 구현
- PostRestController, PostDto, WriterDto 생성

midasWorld committed:
refactor: Post 생성자 - 유효성 체크 추가

midasWorld committed:
feat: Post - 등록, 수정, ID 조회 기능 구현

midasWorld committed:
style: 불필요한 공백 제거

midasWorld committed:
refactor: QueryDSL 제거 → Criteria로 구현

Jo-GyuHyeon commented on src/main/java/org/prgrms/board/config/support/SimpleOffsetPageableHandlerMethodArgumentResolver.java:
```diff
@@ -0,0 +1,57 @@
+import org.springframework.web.method.support.HandlerMethodArgumentResolver;
+import org.springframework.web.method.support.ModelAndViewContainer;
+
+public class SimpleOffsetPageableHandlerMethodArgumentResolver implements HandlerMethodArgumentResolver {
+
+	private static final long DEFAULT_OFFSET = 0L;
+	private static final int DEFAULT_LIMIT = 5;
+
+	private final String offsetParam;
+	private final String limitParam;
```
Bean Validation 으로 간단히 처리 할 수 있었을텐데 해당 방법으로 처리한 이유는 무엇인가요??

midasWorld replied:
Pageable에서 필요없는 필드 말고 필요한 필드만 사용하고, 
좀 더 커스텀해서 사용해 볼수 있지 않을까? 라는 생각으로 시작하게 되었습니다.

Jo-GyuHyeon commented on src/main/java/org/prgrms/board/domain/post/Post.java:
```diff
@@ -0,0 +1,86 @@
+import org.prgrms.board.domain.BaseEntity;
+import org.prgrms.board.domain.user.User;
+
+@Table(name = "post")
```
해당 에노테이션을 사용하지 않아도 되는 방법은 없을까요?

midasWorld replied:
엔티티명과 다른 경우에만 사용하고 그 외에는 생략하는것 말고는 잘 모르겠네요. 😓

Jo-GyuHyeon commented on src/main/java/org/prgrms/board/service/post/PostService.java:
```diff
@@ -0,0 +1,70 @@
+		checkArgument(id != null, "postId must be provided.");
+
+		return postRepository.findById(id)
+			.orElseThrow(() -> new IllegalArgumentException("Could not found post with postId=" + id));
```
IllegalArgumentException 예외라고 하면 올바른 인자값이 아닐경우라고 생각하게 됩니다. 파라미터 타입과 요구하는 조건이 올바르기 때문에 해당 예외는 오해를 불러 일으킬수 있다고 생각이 듭니다.

midasWorld replied:
아.. 대용으로 써도 괜찮다고 생각을 했었는데... 그렇군요. 감사합니다! 바로 적용해보겠습니다. 👍

Jo-GyuHyeon commented on src/main/java/org/prgrms/board/service/post/PostService.java:
```diff
@@ -0,0 +1,70 @@
+		this.userService = userService;
+	}
+
+	public List<Post> findAll(long offset, int limit) {
+		return postQueryRepository.findAll(offset, limit);
```
offset 은 long 타입으로 준 이유는 무엇인가요?

midasWorld replied:
사실 querydsl이 long 타입으로 되어있어서 그랬지만...
offset 이라는게 결국 어디서부터 시작하겠다 인데, 데이터가 수없이 많아질 경우를 대비해서 범위가 넓은 long 타입으로 했다고 이해를 했습니다.

Jo-GyuHyeon commented on src/main/java/org/prgrms/board/service/post/PostService.java:
```diff
@@ -0,0 +1,70 @@
+		checkArgument(isNotEmpty(title), "title must be provided.");
+		checkArgument(isNotEmpty(content), "content must be provided.");
+
+		Post post = postRepository.findById(id)
+			.orElseThrow(() -> new IllegalArgumentException("Could not found post with postId=" + id));
+
+		post.modifyTitleAndContent(title, content);
+
+		return postRepository.save(post);
```
조금더 optional 을 특징을 잘 살려 보는 방법은 어떨까요? 
(체이닝)

Jo-GyuHyeon reviewed (commented):
고생하셨습니다. QueryDSL 사용하기전에 Critera 를 먼저 사용해보시길 권유 드립니다. 👍

Onigiriiiiiiii commented on src/main/java/org/prgrms/board/config/support/SimpleOffsetPageRequest.java:
```diff
@@ -0,0 +1,42 @@
+import org.apache.commons.lang3.builder.ToStringBuilder;
+import org.apache.commons.lang3.builder.ToStringStyle;
+
+public class SimpleOffsetPageRequest implements Pageable {
```
상속을 받아야 하는 이유가 어떻게 되시나 🐉 ?

midasWorld replied:
Pageable이라는 곳에 사용하지 않는 필드들도 많고.. 
그리고 페이징을 하고 다시 Page로 래핑할때 번거롭다는 생각을 하게 되서..
필요한 부분들을 한 클래스에서 그냥 커스텀해서 사용하면 어떨까? 라는 생각으로 시작하게 된것 같네요.

Jo-GyuHyeon replied:
기존에 Pageable 과 네이밍이 겹쳐서 import 할때 문제가 생길 수 있을것 같은데 어떻게 생각하시나요?

midasWorld replied:
흠.. 그러네요.. 네이밍을 좀 고민해보겠습니다.

Onigiriiiiiiii commented on src/main/java/org/prgrms/board/service/post/PostService.java:
```diff
@@ -0,0 +1,70 @@
+
+	@Transactional
+	public Post write(String title, String content, Long writerId) {
+		checkArgument(isNotEmpty(title), "title must be provided.");
```
서비스에서 argument의 여부를 판단하신 이유가 어떻게 되나 🐉 ?

midasWorld replied:
도메인이 커지면 커질 수록 Service안에 다른 Service로 엮이는 일이 많다고 생각을 해서.. 
유효성 검사도 Service에서 하는걸로 초점을 맞췄습니다!

Onigiriiiiiiii commented on src/main/java/org/prgrms/board/service/post/PostService.java:
```diff
@@ -0,0 +1,70 @@
+		Post post = postRepository.findById(id)
+			.orElseThrow(() -> new IllegalArgumentException("Could not found post with postId=" + id));
+
+		post.modifyTitleAndContent(title, content);
```
현재로서는 칼럼이 단 2개여서 부담이 없겠지만 엔티티 스펙변경에 의해 5가지의 컬럼이 변경될 여지가 있을 때 entity 끼리 통신하는 방법이면 좋을 것 같슴둥 💯

Onigiriiiiiiii commented on pom.xml:
```diff
@@ -0,0 +1,109 @@
+            <version>${guava.version}</version>
+        </dependency>
+        <dependency>
+            <groupId>commons-io</groupId>
```
io 라이브러리는 어디에 쓰이는 건가 🐉 ?

midasWorld replied:
아 이게, 파일 관련 라이브러리 인데.. lang3 처럼 유효성검사 편하게 쓰려고 추가를 했습니다.
왠만한건 다 프로필 사진처럼 파일 들어가는 경우가 많아서 기본적으로 추가하게 되네요.

Onigiriiiiiiii commented on src/main/java/org/prgrms/board/domain/post/PostQueryRepository.java:
```diff
@@ -0,0 +1,31 @@
+		this.queryFactory = queryFactory;
+	}
+
+	public List<Post> findAll(long offset, int limit) {
```
jpql로 충분히 커버가 가능해 보이는데 querydsl을 사용하신 이유가 어떻게 되나 🐉  ?

midasWorld replied:
일단 가독성 면도 있구요.
그리고 제가 페이징과 count쿼리를 따로 메서드로 만들었는데.. 이게 결국 성능개선을 위한 것이니....
추후에 조회 컬럼수도 제한해서 DTO로 받을 요량으로 사용하게 되었습니다.

Onigiriiiiiiii commented on src/main/java/org/prgrms/board/web/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,32 @@
+	@ExceptionHandler({IllegalStateException.class, IllegalArgumentException.class})
+	public ResponseEntity<ApiResponse<?>> handlerBadRequestException(Exception e) {
+		log.debug("Bad request exception occurred: {}", e.getMessage(), e);
+		return newResponse(e, HttpStatus.BAD_REQUEST);
```
return 사이에 계행이 있으면 더 좋을 것 같슴둥~

Onigiriiiiiiii reviewed (approved):
여러모로 개인적으로 궁금했던 부분들 위주로 작성했네 🐉  ... 
텍스트로 적기에 양이 많다면 언제든 구두로 요청해서 애기 나눠도 좋을 것 같슴다!

수고하셨습니 🥕 !

midasWorld committed:
fix: PostRepositoryCustomImpl - implements PostRepositoryCustom 추가

midasWorld committed:
refactor: PostRepository → criteria 사용된 query도 여기서 사용되도록 구조 변경 및 리팩토링
- PostQueryRepository → PostRepositoryCustomImpl 클래스 명 변경
- PostRepositoryCustom 인터페이스 생성 → PostRepository에서 상속 되도록 추가
- PostService 수정 기능 fluent 하도록 코드 리팩토링

midasWorld committed:
refactor: PostRepository → criteria 사용된 query도 여기서 사용되도록 구조 변경 및 리팩토링
- PostQueryRepository → PostRepositoryCustomImpl 클래스 명 변경
- PostRepositoryCustom 인터페이스 생성 → PostRepository에서 상속 되도록 추가
- PostService 수정 기능 fluent 하도록 코드 리팩토링

midasWorld committed:
style: GlobalExceptionHandler - return 구문위에 개행 추가.

midasWorld committed:
Merge remote-tracking branch 'origin/2p/#2th_midas' into 2p/#2th_midas

midasWorld committed:
feat: NotFoundException 생성 및 적용
- ServiceRuntimeException 생성 : 커스텀 Exception 추상 클래스
- MessageSourceAccessor 설정 추가 및 MessageUtils 생성 : messages.properties 에서 메시지 관리할 수 있도록
- PostService, UserService, GlobalExceptionHandler 에 적용

Jo-GyuHyeon commented on src/main/java/org/prgrms/board/domain/post/PostRepositoryCustomImpl.java:
```diff
@@ -0,0 +1,39 @@
+
+		Order idDesc = cb.desc(post.get("id"));
+
+		cq.orderBy(idDesc);
```
cq 를 체이닝으로 작성하여 하나의 sql 문 처럼 보이도록 작성해보시면 더욱 가독성이 높아질것 같습니다. 
변수명도 의미있게 단어로 작성해보시는건 어떨까요?

midasWorld replied:
변수명은 확실히 간과했군요.. 😱
빌더는 조금만 더 고민해보겠습니다...

Jo-GyuHyeon commented on src/main/java/org/prgrms/board/web/post/PostRestController.java:
```diff
@@ -0,0 +1,70 @@
+		long offset = pageable.offset();
+		int limit = pageable.limit();
+
+		List<PostDto> posts = postService.findAll(offset, limit).stream()
+			.map(PostDto::new).toList();
```
체이닝이 많아 진다면 한라인당 하나의 체이닝으로 할 수 있도록 하는 것이 일반적인걸로 알고 있습니다.

midasWorld replied:
넵! 적용하겠습니다.
